core/src/impl/Kokkos_Profiling.cpp

```diff
@@ -52,161 +54,138 @@
 #include <chrono>
 
 #if defined(KOKKOS_ENABLE_PROFILING)
-#include <impl/Kokkos_Profiling.hpp>
 #include <cstring>
 #include <unordered_map>
 #include <unordered_set>
 #include <algorithm>
 namespace Kokkos {
 namespace Profiling {
 
-static initFunction initProfileLibrary         = nullptr;
-static finalizeFunction finalizeProfileLibrary = nullptr;
-
-static beginFunction beginForCallback    = nullptr;
-static beginFunction beginScanCallback   = nullptr;
-static beginFunction beginReduceCallback = nullptr;
-static endFunction endForCallback        = nullptr;
-static endFunction endScanCallback       = nullptr;
-static endFunction endReduceCallback     = nullptr;
-
-static pushFunction pushRegionCallback = nullptr;
-static popFunction popRegionCallback   = nullptr;
-
-static allocateDataFunction allocateDataCallback     = nullptr;
-static deallocateDataFunction deallocateDataCallback = nullptr;
-
-static beginDeepCopyFunction beginDeepCopyCallback = nullptr;
-static endDeepCopyFunction endDeepCopyCallback     = nullptr;
-
-static createProfileSectionFunction createSectionCallback   = nullptr;
-static startProfileSectionFunction startSectionCallback     = nullptr;
-static stopProfileSectionFunction stopSectionCallback       = nullptr;
-static destroyProfileSectionFunction destroySectionCallback = nullptr;
-
-static profileEventFunction profileEventCallback = nullptr;
-}  // end namespace Profiling
-
-namespace Profiling {
+static EventSet current_callbacks;
+static EventSet backup_callbacks;
 
-bool profileLibraryLoaded() { return (initProfileLibrary != nullptr); }
+bool profileLibraryLoaded() { return (current_callbacks.init != nullptr); }
```

Comment:
So you need to set this to something to say that you are using anything in the profiling interface? Does it make more sense to check that any of them is set to something else than `nullptr`?

Reply:
That's a really interesting point, and another reason @dalg24 was right about doing this with EventSets. That should just compare `current_callbacks` to `no_profiling`

Reply:
Oh wait, this isn't C, so that doesn't work

Reply:
This bears some thinking about, I'll push a fix soon